File: lytecache-go/lock.go
package lytecache

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"time"
)

// lockKeyPrefix namespaces distributed-lock keys away from ordinary
// user keys within the same cache namespace.
const lockKeyPrefix = "__lock__:"

// lockPollInterval is how often Lock retries acquisition while waiting.
const lockPollInterval = 50 * time.Millisecond

// Lock is a process-safe distributed lock obtained from [Cache.Lock].
type Lock struct {
	cache *Cache
	name  string
	token string
}

// Lock attempts to acquire a process-safe lock named name, polling roughly
// every 50ms until timeout elapses. It is built on the same atomic Add
// (Redis "SET NX") semantics as everything else, so only one holder --
// across goroutines *and* processes sharing the cache file -- can hold a
// given lock name at once.
//
// The lock is also given timeout as its own TTL, as a safety net: if the
// holder's process crashes before calling Release, the lock still expires
// on its own instead of wedging forever.
func (c *Cache) Lock(name string, timeout time.Duration) (*Lock, error) {
	token, err := randomToken()
	if err != nil {
		return nil, fmt.Errorf("lytecache: lock %q: %w", name, err)
	}
	lockKey := lockKeyPrefix + name
	deadline := time.Now().Add(timeout)

	for {
		ok, err := c.Add(lockKey, token, TTL(timeout))
		if err != nil {
			return nil, fmt.Errorf("lytecache: lock %q: %w", name, err)
		}
		if ok {
			return &Lock{cache: c, name: name, token: token}, nil
		}
		if !time.Now().Before(deadline) {
			return nil, fmt.Errorf("%w: could not acquire lock %q within %s", ErrLockTimeout, name, timeout)
		}
		time.Sleep(lockPollInterval)
	}
}

// Release releases the lock. It only deletes the underlying row if this
// Lock's token still matches what is stored -- guarding against releasing
// a lock that expired and was subsequently acquired by someone else.
func (l *Lock) Release() error {
	lockKey := lockKeyPrefix + l.name
	data, _, err := encodeValue(l.token)
	if err != nil {
		return fmt.Errorf("lytecache: release lock %q: %w", l.name, err)
	}
	_, err = l.cache.writeDB.Exec(
		`DELETE FROM cache WHERE namespace = ? AND key = ? AND value = ?`,
		l.cache.namespace, lockKey, data)
	if err != nil {
		return fmt.Errorf("lytecache: release lock %q: %w", l.name, err)
	}
	return nil
}

func randomToken() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}

File: lytecache-node/src/lock.ts
import { randomUUID } from "node:crypto";
import "./dispose-polyfill.js";
import { LockTimeoutError } from "./errors.js";
import { sleepSyncMs } from "./sleep-sync.js";

/** The minimal surface {@link CacheLock} needs from {@link LyteCache}, to avoid a circular import. */
export interface LockHost {
  tryAcquireLock(lockKey: string, token: string, ttlSeconds: number): boolean;
  releaseLock(lockKey: string, token: string): boolean;
}

export interface LockOptions {
  /** How long to wait to acquire before throwing {@link LockTimeoutError}. Default 30000. */
  timeoutMs?: number;
  /** How often to retry while waiting. Default 50. */
  pollMs?: number;
}

/**
 * A process-safe distributed lock built on the cache's atomic {@link LyteCache.add} semantics:
 * acquiring is a `SET NX` on a `"lock:" + name` key, so at most one holder -- across threads and
 * processes -- can hold it at a time.
 *
 * The lock's TTL equals the acquisition timeout: if the holder crashes without releasing, the
 * lock expires and becomes available again rather than being stuck forever. Each acquisition
 * writes a unique random token as the lock's value, so release only ever removes a lock this
 * instance still holds -- never one that already expired and was re-acquired by someone else.
 *
 * Supports `using` (TC39 explicit resource management) via `Symbol.dispose`.
 */
export class CacheLock {
  private readonly host: LockHost;
  private readonly lockKey: string;
  private readonly token = randomUUID();
  private readonly ttlSeconds: number;
  private released = false;

  constructor(host: LockHost, name: string, options: LockOptions = {}) {
    this.host = host;
    this.lockKey = `lock:${name}`;
    const timeoutMs = options.timeoutMs ?? 30_000;
    const pollMs = options.pollMs ?? 50;
    this.ttlSeconds = timeoutMs / 1000;
    this.acquire(timeoutMs, pollMs);
  }

  private acquire(timeoutMs: number, pollMs: number): void {
    const deadline = Date.now() + timeoutMs;
    for (;;) {
      if (this.host.tryAcquireLock(this.lockKey, this.token, this.ttlSeconds)) {
        return;
      }
      if (Date.now() >= deadline) {
        throw new LockTimeoutError(
          `could not acquire lock "${this.lockKey.slice(5)}" within ${timeoutMs}ms`,
        );
      }
      sleepSyncMs(pollMs);
    }
  }

  /** Releases the lock, if this instance still holds it. Safe to call more than once. */
  release(): void {
    if (!this.released) {
      this.released = true;
      this.host.releaseLock(this.lockKey, this.token);
    }
  }

  [Symbol.dispose](): void {
    this.release();
  }
}
